cvat-ui/src/containers/annotation-page/top-bar/top-bar.tsx

```diff
@@ -107,30 +132,80 @@ function mapDispatchToProps(dispatch: any): DispatchToProps {
 
 type Props = StateToProps & DispatchToProps;
 class AnnotationTopBarContainer extends React.PureComponent<Props> {
+    private autoSaveInterval: number | undefined;
+
     public componentDidUpdate(): void {
         const {
             jobInstance,
             frameNumber,
+            frameChangeTime,
+            frameSpeed,
             playing,
             canvasIsReady,
-            onChangeFrame,
             onSwitchPlay,
         } = this.props;
 
+        
         if (playing && canvasIsReady) {
             if (frameNumber < jobInstance.stopFrame) {
+                const delay: number = frameChangeTime ? Math.max(0,~~(1000/frameSpeed) -  new Date().getTime() + frameChangeTime) : 0;
                 setTimeout(() => {
                     const { playing: stillPlaying } = this.props;
                     if (stillPlaying) {
-                        onChangeFrame(frameNumber + 1);
+                        this.onChangeFrame(frameNumber + 1, new Date().getTime());
                     }
-                });
+                }, delay);
             } else {
                 onSwitchPlay(false);
             }
         }
     }
 
+    componentDidMount(): void {
+        const {
+            autoSave, 
+            autoSaveInterval,
+            saving,
+            unsaved,
+         } = this.props;
+
+        this.autoSaveInterval = window.setInterval((autoSave: boolean, saving: boolean): void => {
+            if (autoSave && !saving) {
+                this.onSaveAnnotation();
+            }
+        }, autoSaveInterval, autoSave, saving);
+
+        if (unsaved) {
+            window.addEventListener('beforeunload', this.beforeUnloadCallback);
+        }
+    }
+
+    public componentWillUnmount() : void {
+        window.clearInterval(this.autoSaveInterval);
+        window.removeEventListener('beforeunload', this.beforeUnloadCallback);
+    }
+
+    private beforeUnloadCallback(event: BeforeUnloadEvent): any {
+        const confirmationMessage = 'You have unsaved changes, please confirm leaving this page.';
+        (event || window.event).returnValue = confirmationMessage;
+        return confirmationMessage;
+    }
+
+    private onChangeFrame(newFrame: number, time: number | null = null): void {
+        const { 
+            canvasInstance,
+            canvasIsReady,
+            onChangeFrame,
+            resetZoom,
+        } = this.props;
+
+        onChangeFrame(newFrame, time);
+
+        if (canvasIsReady && resetZoom) {
+            canvasInstance.fit();
+        }
```

Comment:
It will be executed immediately after ``onChangeFrame()`` function (which is async). So, frame isn't expected to be changed yet. 

To make fit() I would recommend you to use ``canvas.setup`` event (see canvas-wrapper component, initialSetup method)